### PR TITLE
feature(pu): add on_policy_distillation

### PR DESCRIPTION
**Pull request by puyuan1996**
- add init version of on_policy_distillation

### CODE REVIEW

File: lightrft/trainer/advantage_calculator.py [hunk 1]
```diff
@@ -717,6 +718,160 @@ def compute(
...
+    reverse_kl = torch.clamp(reverse_kl, min=-20.0, max=20.0)
+
+    opd_adv = -opd_kl_coef * reverse_kl
+
+    # Rely solely on action_mask for padding filtering.
+    # Previous code also masked teacher_log_probs == 0.0, but log_prob=0 is a
+    # legitimate value (P=1). With the padding-strip fix in _fetch_teacher_logprobs,
+    # teacher logprobs are now correctly aligned and action_mask is sufficient.
+    if action_mask is not None:
+        opd_adv = opd_adv * action_mask
```
Comment (PaParaZz1) [inline lightrft/trainer/advantage_calculator.py]: `*=` or `mul_`

File: lightrft/trainer/experience_maker.py [hunk 2]
```diff
@@ -400,6 +400,15 @@ def make_experience_list(self, all_prompts: Union[str, List[str]], **generate_kw
                     generate_kwargs["gamma"],
                 )
                 experience.advantages = deepcopy(experience.returns)
+            elif self.advantage_estimator in ("on_policy_distillation", "on_policy_distillation_hybrid"):
```
Comment (PaParaZz1) [inline lightrft/trainer/experience_maker.py]: why not merge this branch into the one of line 396

File: lightrft/trainer/fast_exp_maker.py [hunk 3]
```diff
@@ -58,6 +58,17 @@
 from .image_utils import normalize_images, get_images_num
 from .video_utils import normalize_videos, get_videos_num
 
+# On-Policy Distillation imports
+try:
```
Comment (PaParaZz1) [inline lightrft/trainer/fast_exp_maker.py]: why use try-except here, we always enable OPD

File: lightrft/trainer/fast_exp_maker.py [hunk 4]
```diff
@@ -441,7 +452,13 @@ def __init__(
...
         """
         self.reward_model = reward_model
-        self.remote_rm_url = remote_rm_url
+        # Ensure remote_rm_url is a list for consistent iteration
+        if remote_rm_url is None:
+            self.remote_rm_url = None
+        elif isinstance(remote_rm_url, str):
+            self.remote_rm_url = [remote_rm_url]
+        else:
+            self.remote_rm_url = list(remote_rm_url)
```
Comment (PaParaZz1) [inline lightrft/trainer/fast_exp_maker.py:455]: raise error for else branch

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: lightrft/trainer/spmd_ppo_trainer.py [hunk 5]
```diff
@@ -422,6 +427,22 @@ def ppo_train(self, global_steps=0):  # Currently using this rewritten ppo_train
...
+                # Collect reverse KL from all experiences
+                if isinstance(all_opd_reverse_kl[0], torch.Tensor):
+                    opd_kl_tensor = torch.cat([t.to(device).float() for t in all_opd_reverse_kl])
+                else:
+                    opd_kl_tensor = torch.tensor(all_opd_reverse_kl, dtype=torch.float32, device=device)
+                # Mask out zero values (padding)
+                non_zero_mask = opd_kl_tensor != 0
+                if non_zero_mask.any():
+                    masked_kl = opd_kl_tensor[non_zero_mask]
+                    status_mean["opd_reverse_kl_mean"] = masked_kl.mean().item()
```
Comment (PaParaZz1) [inline lightrft/trainer/spmd_ppo_trainer.py:440]: here only mean and max are enough

File: examples/on_policy_distillation/test_opd.py [hunk 6]
```diff
@@ -0,0 +1,453 @@
+"""
```
Comment (PaParaZz1) [inline examples/on_policy_distillation/test_opd.py:1]: use `pytest` to write unittest

File: lightrft/trainer/fast_exp_maker.py (hunk 4 above)
Comment (PaParaZz1) [reply lightrft/trainer/fast_exp_maker.py:455]: don't reuse `remote_rm_url`, add a new `teacher_model_url` argument

File: lightrft/trainer/fast_exp_maker.py [hunk 7]
```diff
@@ -1414,6 +1446,92 @@ def _process_multi_image_video_thws(
...
+            # sequences is [prompt, response, eos, pad, pad, ...] — the padding
+            # tokens would cause SGLang to return logprobs for pad positions,
+            # making the [-resp_len:] slice grab wrong tokens.
+            input_ids_list = []
+            for j in range(sequences.shape[0]):
+                valid_len = int(attention_mask[j].sum().item())
+                input_ids_list.append(sequences[j, :valid_len].cpu().tolist())
+
+            try:
+                loop = asyncio.new_event_loop()
```
Comment (PaParaZz1) [inline lightrft/trainer/fast_exp_maker.py:1505]: why do we need asyncio here, all the operations in `fast_exp_maker` are sync version

File: examples/on_policy_distillation/README_zh.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+- 学生无法超越教师的能力
+- 需要足够的计算资源进行教师推理
+
+## 文件结构
+
+```
+examples/on_policy_distillation/
+├── README.md                           # 英文文档
+├── README_zh.md                        # 本文件
+├── run_opd_qwen_2.sh                   # 训练脚本
```
Comment (PaParaZz1) [inline examples/on_policy_distillation/README_zh.md]: name is mismatch

File: examples/on_policy_distillation/README_zh.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+| 方法 | 奖励信号 | 模式 | 需要 RM |
+|------|---------|------|---------|
+| GRPO | 任务特定奖励 | 在线 | 是 |
+| DPO | 偏好对 | 离线 | 否 |
+| **OPD** | 教师对数概率 | 在线 | 否（使用教师） |
+
+### 优势
+
+- 无需单独训练奖励模型
```
Comment (PaParaZz1) [inline examples/on_policy_distillation/README_zh.md]: 讲清楚一点吧，和 outcome reward 之间的关系

File: examples/on_policy_distillation/README_zh.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        return advantages, returns, {"opd_reverse_kl": reverse_kl}
+```
+
+### 2. 教师对数概率获取器
+
+**文件**: `examples/on_policy_distillation/on_policy_distillation_reward.py`
+
+- 异步 HTTP 请求到教师服务器
+- 支持 SGLang 和 vLLM 响应格式
+- 自动重试（指数退避）
```
Comment (PaParaZz1) [inline examples/on_policy_distillation/README_zh.md]: 指数退避是啥意思

File: examples/on_policy_distillation/README_zh.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
+# LightRFT 在线策略蒸馏 (OPD)
```
Comment (PaParaZz1) [inline examples/on_policy_distillation/README_zh.md:1]: OPD 的 ref 链接放一下

File: lightrft/trainer/advantage_calculator.py [hunk 12]
```diff
@@ -717,6 +718,160 @@ def compute(
...
+    student_log_probs: torch.Tensor,
+    teacher_log_probs: torch.Tensor,
+    action_mask: Optional[torch.Tensor],
+    opd_kl_coef: float,
+) -> Tuple[torch.Tensor, Dict]:
+    """
+    Compute OPD reverse KL penalty: -opd_kl_coef * (student_logp - teacher_logp).
+
+    Shared helper for both pure and hybrid OPD modes.
+    Aligned with Slime's apply_opd_kl_to_advantages.
```
Comment (PaParaZz1) [inline lightrft/trainer/advantage_calculator.py]: remove this line

File: lightrft/trainer/experience_maker.py [hunk 13]
```diff
@@ -575,6 +584,83 @@ def process_experiences(self, experiences: List[Experience]) -> Tuple[List[Exper
         """
         args = self.strategy.args
 
+        # On-policy distillation: query teacher model for log probs, then use GRPO reward shaping
+        if args.advantage_estimator in ("on_policy_distillation", "on_policy_distillation_hybrid"):
+            if self.remote_rm_url is None or len(self.remote_rm_url) == 0:
```
Comment (PaParaZz1) [inline lightrft/trainer/experience_maker.py]: simplify this part, `fast_exp_maker.py` also add similar implementation, maybe you can abstract a common function used for both sides

File: lightrft/trainer/experience_maker.py [hunk 14]
```diff
@@ -575,6 +584,83 @@ def process_experiences(self, experiences: List[Experience]) -> Tuple[List[Exper
...
+                # Pure distillation: zero rewards, learning signal from OPD KL only
+                zero_rewards = torch.zeros(sum(exp.sequences.size(0) for exp in experiences))
+                rewards = zero_rewards.chunk(len(experiences))
+                return experiences, list(rewards)
+            else:
+                # Hybrid: use task rewards with GRPO normalization
+                rewards = torch.cat([experience.info["reward"] for experience in experiences])
+                rewards = rewards.reshape(-1, args.n_samples_per_prompt)
+                baseline = rewards.mean(-1, keepdim=True)
+                rewards = (rewards - baseline) / (rewards.std(-1, keepdim=True) + 1e-9)
```
Comment (PaParaZz1) [inline lightrft/trainer/experience_maker.py]: should there be controlled by some unified reward norm config?

File: lightrft/trainer/fast_exp_maker.py [hunk 15]
```diff
@@ -1414,6 +1446,92 @@ def _process_multi_image_video_thws(
...
+                    loop.close()
+
+                # Align teacher log probs to action_log_probs shape [batch_size, num_actions].
+                # Use action_mask indices directly — works regardless of left/right padding.
+                batch_size = sequences.shape[0]
+                aligned_teacher_lp = torch.zeros(batch_size, num_actions, dtype=torch.float32)
+                for i, (tlp, resp_len) in enumerate(zip(teacher_lp_list, response_lengths)):
+                    if resp_len == 0:
+                        continue
+                    valid_indices = torch.where(action_mask[i] == 1)[0]
```
Comment (PaParaZz1) [inline lightrft/trainer/fast_exp_maker.py:1531]: check action_mask necessity here

File: lightrft/trainer/opd_utils.py [hunk 16]
```diff
@@ -0,0 +1,258 @@
+"""
```
Comment (PaParaZz1) [inline lightrft/trainer/opd_utils.py:1]: this file is not used at all

File: lightrft/trainer/advantage_calculator.py [hunk 17]
```diff
@@ -717,6 +718,160 @@ def compute(
...
+
+    info_dict = {}
+    if action_mask is not None:
+        masked_rkl = reverse_kl * action_mask
+        info_dict["opd_reverse_kl"] = masked_rkl.sum(-1) / action_mask.sum(-1).clamp(min=1)
+
+    return opd_adv, info_dict
+
+
+def _whiten_advantages(advantages: torch.Tensor, action_mask: Optional[torch.Tensor], eps: float = 1e-8) -> torch.Tensor:
```
Comment (PaParaZz1) [inline lightrft/trainer/advantage_calculator.py]: this func is not used

File: lightrft/trainer/advantage_calculator.py [hunk 18]
```diff
@@ -791,21 +951,42 @@ def normalize_advantages_cross_batch(experiences: List, advantage_estimator: str
 
     # Concatenate into vectors
     advantages_vector = torch.cat(all_advantages, dim=0).float()
-    action_masks_vector = torch.cat(all_action_masks, dim=0)
-    num_actions = action_masks_vector.sum()
+    action_masks_vector = torch.cat(all_action_masks, dim=0).float()
```
Comment (PaParaZz1) [inline lightrft/trainer/advantage_calculator.py]: why add `vector` suffix, simplify it

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: lightrft/trainer/opd_utils.py (hunk 16 above)
Comment (puyuan1996) [reply lightrft/trainer/opd_utils.py:1]: used in experience_maker.py and fast_experience_maker.py

File: examples/on_policy_distillation/README_zh.md (hunk 10 above)
Comment (puyuan1996) [reply examples/on_policy_distillation/README_zh.md]: 是指重试延迟在每次尝试后会翻倍（例如 1s → 2s → 4s），这种机制能有效防止在瞬态故障期间导致服务器过载。readme中更新了哈

File: lightrft/trainer/fast_exp_maker.py (hunk 7 above)
Comment (puyuan1996) [reply lightrft/trainer/fast_exp_maker.py:1505]: 虽然 fast_exp_maker 的主流程是同步的（主要处理本地 GPU 计算），但 get_teacher_logprobs_by_ids 涉及向远程 Teacher 模型发送 HTTP 请求。
如果使用同步请求，Batch 内的样本只能串行请求，网络 I/O 延迟会严重阻塞整个经验收集过程。因此，内部使用了异步并发（如 aiohttp + asyncio.gather）来同时发送整个 Batch 的请求。这里手动创建 Event Loop 和 run_until_complete，是为了在不改变外部同步架构的前提下，桥接并驱动局部的异步高并发网络请求，从而大幅降低通信耗时。

File: lightrft/trainer/fast_exp_maker.py (hunk 15 above)
Comment (puyuan1996) [reply lightrft/trainer/fast_exp_maker.py:1531]: 确认过了，这里应该没有问题。[experience_maker.py](https://github.com/opendilab/LightRFT/pull/43/files/9aa39a99b4f96aee1abd355a42b07141074ba876#diff-1635836ebf4f622236f632a14b31cc32d9f111a9e42468f63923b20366296370)里面的有问题，但是 experience_maker.py是没有用到的，我把experience_maker.py里面opd相关实现部分删除了哈

File: lightrft/trainer/experience_maker.py (hunk 13 above)
Comment (puyuan1996) [reply lightrft/trainer/experience_maker.py]: 我把experience_maker.py里面opd相关实现部分删除了哈，因为没有用到。

File: lightrft/trainer/fast_exp_maker.py (hunk 7 above)
Comment (PaParaZz1) [reply lightrft/trainer/fast_exp_maker.py:1505]: 那这里写成专门一个工具函数吧，remote_rm_url, teacher_url 都可以复用类似的功能，且加上一行简单的注释

File: examples/on_policy_distillation/README_zh.md (hunk 11 above)
Comment (PaParaZz1) [reply examples/on_policy_distillation/README_zh.md:1]: 这个还是在文档最开始就有个可以点的链接吧

File: examples/gsm8k_geo3k/reward_models_utils.py [hunk 19]
```diff
@@ -383,6 +384,10 @@ def mix_rewards(
 
         final_reward[i] = r
 
+    # When use_task_reward=False, zero out final rewards but keep metrics for logging
```
Comment (PaParaZz1) [inline examples/gsm8k_geo3k/reward_models_utils.py:387]: this part should be implemented at the beginning of this function

File: lightrft/trainer/opd_utils.py (hunk 16 above)
Comment (PaParaZz1) [reply lightrft/trainer/opd_utils.py:1]: then this module should be placed inside of `lightrft`

Comment (PaParaZz1) [reply lightrft/trainer/opd_utils.py:1]: should be named into `onpolicy_distillation_utils`

**Review by PaParaZz1 — APPROVED** [review]

File: examples/on_policy_distillation/run_opd_qwen.sh [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+    --pretrain "$STUDENT_MODEL_PATH" \
+    --save_trajectories \
+    --advantage_estimator "${ADVANTAGE_ESTIMATOR}" \
+    --opd_kl_coef ${OPD_KL_COEF} \
+    --fsdp \
+    --use_kl_loss \
+    --flash_attn \
+    --engine_type sglang \
+    --enable_engine_sleep \
+    --rm_use_engine \
```
Comment (PaParaZz1) [inline examples/on_policy_distillation/run_opd_qwen.sh]: simplify the args and remove unused fields, such `rm_use_engine`

File: examples/gsm8k_geo3k/reward_models_utils.py (hunk 19 above)
Comment (puyuan1996) [reply examples/gsm8k_geo3k/reward_models_utils.py:387]: 为什么呢 因为format acc reward必须记录，否则没法判断回答的情况，所以前面先算出在后面根据use_task_reward来决定是否设置为0

File: lightrft/trainer/spmd_ppo_trainer.py (hunk 5 above)
Comment (puyuan1996) [reply lightrft/trainer/spmd_ppo_trainer.py:440]: 目前暂时都保留了哈
